File: ORMConvertor/JavaTests/src/test/java/cz/stochel/ormconvertor/javatests/differential/DifferentialMutation.java
package cz.stochel.ormconvertor.javatests.differential;

import java.util.ArrayList;
import java.util.List;
import java.util.function.UnaryOperator;

/**
 * The deliberately wrong translations F13 asks to be detected (decision 089). Each one is
 * applied to the <em>generated artifact</em> and not to the translator: mutating the
 * translator would be a tool of its own, with bugs of its own, and most of its mutations
 * would produce artifacts that do not even compile - which proves nothing about comparing
 * results. What is under test here is the sensitivity of the comparison, so the mutation
 * has to reach the rows.
 *
 * <p>The counterpart is {@code DifferentialMutation.cs}. The two are written twice rather
 * than shared, as the renderer is - but unlike the renderer, a drift between them cannot
 * make the suites agree wrongly: each side only ever asserts that its own mutation was
 * caught.
 */
public record DifferentialMutation(String key, String name, UnaryOperator<String> apply) {

    public static List<DifferentialMutation> all() {
        List<DifferentialMutation> mutations = new ArrayList<>();
        mutations.add(new DifferentialMutation("filter", "the filter is dropped", DifferentialMutation::dropFilter));
        mutations.add(new DifferentialMutation("operator", "the comparison operator is flipped", DifferentialMutation::flipOperator));
        mutations.add(new DifferentialMutation("ordering", "the ordering is dropped", DifferentialMutation::dropOrdering));
        mutations.add(new DifferentialMutation("rowCount", "the row count is changed", DifferentialMutation::changeRowCount));
        mutations.add(new DifferentialMutation("projection", "two projected fields are swapped", DifferentialMutation::swapProjection));
        return mutations;
    }

    /** The mutation a matrix entry names, or a failure that says the name is not one. */
    public static DifferentialMutation of(String key) {
        return all().stream()
                .filter(mutation -> mutation.key().equals(key))
                .findFirst()
                .orElseThrow(() -> new IllegalStateException(
                        "matrix.txt names the mutation \"" + key + "\", which is none of "
                                + all().stream().map(DifferentialMutation::key).toList() + "."));
    }

    /** Every target writes its filter on a line of its own - WHERE, where, .Where(. */
    private static String dropFilter(String artifact) {
        return lines(artifact, line -> !line.matches("(?s)\\s*(WHERE|where)\\b.*") && !line.contains(".Where("));
    }

    /**
     * The comparison the filter is built on. Whitespace on both sides is what tells it from
     * the angle brackets of {@code List<Product>} and from the arrow of a lambda; without
     * that the mutation rewrites the generics and the artifact does not compile, which would
     * prove that a broken artifact differs rather than that a wrong query does.
     */
    private static String flipOperator(String artifact) {
        // A MyBatis statement lives inside XML, where the comparison is escaped, so both
        // spellings have to be flipped or the mutation would never reach that one target.
        String flipped = artifact.replaceAll("(?<=\\s)&gt;(?=\\s)", "\u0002");
        flipped = flipped.replaceAll("(?<=\\s)&lt;(?=\\s)", "&gt;");
        flipped = flipped.replace("\u0002", "&lt;");

        flipped = flipped.replaceAll("(?<=\\s)>(?=\\s)", "\u0001");
        flipped = flipped.replaceAll("(?<=\\s)<(?=\\s)", ">");
        return flipped.replace('\u0001', '<');
    }

    /** Ordering lives on its own line as well - ORDER BY, order by, .OrderBy (descending included). */
    private static String dropOrdering(String artifact) {
        return lines(artifact, line -> !line.matches("(?s).*(ORDER\\s+BY|order\\s+by)\\b.*") && !line.contains(".OrderBy"));
    }

    /**
     * The row count of a paginated query, wherever the target put it: inside TOP or on the
     * query object. Three of six rows is a proper prefix, so two is a different answer
     * whichever way the target slices.
     */
    private static String changeRowCount(String artifact) {
        return artifact.replaceAll("(TOP\\s*\\(\\s*|setMaxResults\\()3(\\s*\\))", "$12$2");
    }

    /**
     * The two projected expressions swapped under their own aliases: the row keeps its shape
     * and every field holds the other one's value. Over this fixture Sku and ProductName
     * differ in every row, so nothing survives the swap by coincidence.
     */
    private static String swapProjection(String artifact) {
        return artifact.replaceAll(
                "(\\w+\\.)(\\w+)(\\s+(?:AS|as)\\s+)(\\w+)(,\\s*)(\\w+\\.)(\\w+)(\\s+(?:AS|as)\\s+)(\\w+)",
                "$1$7$3$4$5$6$2$8$9");
    }

    /**
     * The artifact without the lines a rule rejects - and the artifact itself when it
     * rejects none. Returning a rebuilt text either way would make every no-op look like a
     * mutation, because rebuilding normalizes the line endings; a query with no ordering
     * would then be reported as having lost one.
     */
    private static String lines(String artifact, java.util.function.Predicate<String> keep) {
        String[] all = artifact.replace("\r\n", "\n").split("\n", -1);
        List<String> kept = new ArrayList<>();

        for (String line : all) {
            if (keep.test(line)) {
                kept.add(line);
            }
        }

        return kept.size() == all.length ? artifact : String.join(System.lineSeparator(), kept);
    }
}

File: ORMConvertor/Tests/Differential/DifferentialMutation.cs
using System.Text.RegularExpressions;

namespace Tests.Differential;

/// <summary>
/// The deliberately wrong translations F13 asks to be detected (decision 089). Each one is
/// applied to the <em>generated artifact</em> and not to the translator: mutating the
/// translator would be a tool of its own, with bugs of its own, and most of its mutations
/// would produce artifacts that do not even compile - which proves nothing about comparing
/// results. What is under test here is the sensitivity of the comparison, so the mutation
/// has to reach the rows.
///
/// The list is the one the decision names, and the fixture of that decision is chosen so
/// that every one of them changes the outcome over it. A mutation that leaves the artifact
/// unchanged is itself a failure: it would pass by accident.
///
/// The transformations are textual and language-blind on purpose. The artifact of every
/// target puts the filter, the ordering and the projection where a line or a comma says
/// they are, so one rule reaches SQL, HQL, JPQL and a LINQ chain alike.
/// </summary>
internal sealed record DifferentialMutation(string Key, string Name, Func<string, string> Apply)
{
    public static IEnumerable<DifferentialMutation> All() =>
    [
        new("filter", "the filter is dropped", DropFilter),
        new("operator", "the comparison operator is flipped", FlipOperator),
        new("ordering", "the ordering is dropped", DropOrdering),
        new("rowCount", "the row count is changed", ChangeRowCount),
        new("projection", "two projected fields are swapped", SwapProjection),
    ];

    /// <summary>The mutation a matrix entry names, or a failure that says the name is not one.</summary>
    public static DifferentialMutation Of(string key)
        => All().SingleOrDefault(mutation => mutation.Key == key)
           ?? throw new InvalidOperationException(
               $"matrix.txt names the mutation \"{key}\", which is none of "
               + string.Join(", ", All().Select(mutation => mutation.Key)) + ".");

    /// <summary>Every target writes its filter on a line of its own - WHERE, where, .Where(.</summary>
    private static string DropFilter(string artifact)
        => Lines(artifact, line => !Regex.IsMatch(line, @"^\s*(WHERE|where)\b") && !line.Contains(".Where("));

    /// <summary>
    /// The comparison the filter is built on. Over this fixture the two directions select
    /// disjoint, non-empty sets, so the flip cannot go unnoticed.
    /// </summary>
    private static string FlipOperator(string artifact)
    {
        // Whitespace on both sides, which is what tells a comparison from the angle
        // brackets of List&lt;Product&gt; and from the arrow of a lambda. Without it the
        // mutation rewrites the generics of the artifact and the thing does not compile -
        // which would prove that a broken artifact differs, not that a wrong query does.
        var flipped = Regex.Replace(artifact, @"(?<=\s)>(?=\s)", "\u0001");
        flipped = Regex.Replace(flipped, @"(?<=\s)<(?=\s)", ">");
        return flipped.Replace('\u0001', '<');
    }

    /// <summary>
    /// Ordering lives on its own line as well - ORDER BY, order by, .OrderBy. The last one
    /// without its parenthesis on purpose: a descending ordering reaches the EF Core
    /// artifact as .OrderByDescending, and a rule that missed it would leave the artifact
    /// untouched for every query that orders the other way.
    /// </summary>
    private static string DropOrdering(string artifact)
        => Lines(artifact, line => !Regex.IsMatch(line, @"(ORDER\s+BY|order\s+by)\b") && !line.Contains(".OrderBy"));

    /// <summary>
    /// The row count of a paginated query, wherever the target put it: inside TOP, in a
    /// Take call, on the query object. Three of six rows is a proper prefix, so two is a
    /// different answer whichever way the target slices.
    /// </summary>
    private static string ChangeRowCount(string artifact)
        => Regex.Replace(artifact, @"(TOP\s*\(\s*|\.Take\(|SetMaxResults\(|setMaxResults\()3(\s*\))", "${1}2${2}");

    /// <summary>
    /// The two projected expressions swapped under their own aliases: the row keeps its
    /// shape and every field holds the other one's value. Over this fixture Sku and
    /// ProductName differ in every row, so nothing survives the swap by coincidence.
    /// </summary>
    private static string SwapProjection(string artifact)
    {
        // SELECT p.A AS X, p.B AS Y  - the shape of SQL, HQL and JPQL.
        var swapped = Regex.Replace(
            artifact,
            @"(\w+\.)(\w+)(\s+(?:AS|as)\s+)(\w+)(,\s*)(\w+\.)(\w+)(\s+(?:AS|as)\s+)(\w+)",
            "$1$7$3$4$5$6$2$8$9");

        if (swapped != artifact)
        {
            return swapped;
        }

        // new { X = p.A, Y = p.B } - the shape of a LINQ projection.
        return Regex.Replace(
            artifact,
            @"(\w+)(\s*=\s*)(\w+\.)(\w+)(,\s*)(\w+)(\s*=\s*)(\w+\.)(\w+)",
            "$1$2$3$9$5$6$7$8$4");
    }

    /// <summary>
    /// The artifact without the lines a rule rejects - and the artifact itself when it
    /// rejects none. Returning a rebuilt text either way would make every no-op look like a
    /// mutation, because rebuilding normalizes the line endings; a query with no ordering
    /// would then be reported as having lost one.
    /// </summary>
    private static string Lines(string artifact, Func<string, bool> keep)
    {
        var lines = artifact.Replace("\r\n", "\n").Split('\n');
        var kept = lines.Where(keep).ToList();

        return kept.Count == lines.Length ? artifact : string.Join(Environment.NewLine, kept);
    }
}
